server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -861,17 +854,38 @@ private void checkKeepAliveLimit(long keepAlive) {
         }
     }
 
-    private void processFailure(ShardSearchRequest request, ReaderContext context, Exception e) {
-        if (context.singleSession() || request.scroll() != null) {
+    private <T> ActionListener<T> wrapFailureListener(ActionListener<T> listener, ReaderContext context, Releasable releasable) {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(T resp) {
+                releasable.close();
+                listener.onResponse(resp);
+            }
+
+            @Override
+            public void onFailure(Exception exc) {
+                processFailure(context, exc);
+                releasable.close();
+                listener.onFailure(exc);
```

Comment:
I know `processFailure` and `markAsUsed` do not throw exceptions, but can we use "Releases.close()` to ensure that we always notify the listener?

Reply:
++, https://github.com/elastic/elasticsearch/pull/61831/commits/16ea240ad54ca1a76da955b934cc91066bdaee11